wire-gson-support/src/main/java/com/squareup/wire/MessageTypeAdapter.kt

```diff
@@ -0,0 +1,161 @@
+      fieldBindings[binding.name] = binding
+    }
+    this.fieldBindings = unmodifiableMap(fieldBindings)
+  }
+
+  @Throws(IOException::class)
+  override fun write(out: JsonWriter, message: M?) {
+    if (message == null) {
+      out.nullValue()
+      return
+    }
+
+    out.beginObject()
+    for (tagBinding in messageAdapter.fieldBindings().values) {
+      val value = tagBinding.get(message) ?: continue
+      out.name(tagBinding.name)
+      emitJson(out, value, tagBinding.singleAdapter(), tagBinding.label)
+    }
+    out.endObject()
+  }
+
+  private fun emitJson(out: JsonWriter, value: Any, adapter: ProtoAdapter<*>, label: Label) {
+    if (adapter === ProtoAdapter.UINT64) {
+      if (label.isRepeated) {
+        val longs = value as List<Long>
+        out.beginArray()
+        val count = longs.size
+        for (i in 0 until count) {
+          emitUint64(longs[i], out)
+        }
+        out.endArray()
+      } else {
+        emitUint64(value as Long, out)
+      }
+    } else {
+      gson.toJson(value, value.javaClass, out)
+    }
+  }
+
+  private fun emitUint64(value: Long, out: JsonWriter) {
+    if (value < 0) {
+      val unsigned = POWER_64.add(BigInteger.valueOf(value))
+      out.value(unsigned)
+    } else {
+      out.value(value)
+    }
+  }
+
+  @Throws(IOException::class)
+  override fun read(`in`: JsonReader): M? {
+    if (`in`.peek() == JsonToken.NULL) {
+      `in`.nextNull()
+      return null
+    }
+
+    val elementAdapter = gson.getAdapter(JsonElement::class.java)
+    val builder = messageAdapter.newBuilder()
+
+    `in`.beginObject()
+    while (`in`.peek() != JsonToken.END_OBJECT) {
+      val name = `in`.nextName()
+
+      val fieldBinding = fieldBindings[name]
+      if (fieldBinding == null) {
+        `in`.skipValue()
+      } else {
+        val element = elementAdapter.read(`in`)
+        val value = parseValue(fieldBinding, element)
+        fieldBinding.set(builder, value)
+      }
+    }
+
+    `in`.endObject()
+    return builder.build()
+  }
+
+  private fun parseValue(fieldBinding: FieldBinding<*, *>, element: JsonElement): Any {
+    if (fieldBinding.label.isRepeated) {
+      if (element.isJsonNull) {
+        return emptyList<Any>()
+      }
+
+      val itemType = fieldBinding.singleAdapter().javaType
+
+      val array = element.asJsonArray
+      val result = ArrayList<Any>(array.size())
+      for (item in array) {
+        result.add(gson.fromJson(item, itemType))
+      }
+      return result
+    }
+
+    if (fieldBinding.isMap()) {
+      if (element.isJsonNull) {
+        return emptyMap<Any, Any>()
+      }
+
+      val keyType = fieldBinding.keyAdapter().javaType
+      val valueType = fieldBinding.singleAdapter().javaType
+
+      val `object` = element.asJsonObject
+      val result = LinkedHashMap<Any, Any>(`object`.size())
+      for (entry in `object`.entrySet()) {
+        val key = gson.fromJson(entry.key, keyType)
```

Comment:
this could also be `map`ish but I'm not sure if there's a way to go directly to a map without an intermediate collection. that's unfortunate...

this also should look up the Gson `TypeAdapter` once instead of using `fromJson` though. should be separate change

Reply:
Fixed.